source/server/drain_manager_impl.cc

```diff
@@ -35,30 +32,25 @@ bool DrainManagerImpl::drainClose() const {
     return false;
   }
 
-  // We use the tick time as in increasing chance that we shutdown connections.
-  return static_cast<uint64_t>(drain_time_completed_.load()) >
-         (server_.random().random() % server_.options().drainTime().count());
-}
-
-void DrainManagerImpl::drainSequenceTick() {
-  ENVOY_LOG(trace, "drain tick #{}", drain_time_completed_.load());
-  ASSERT(drain_time_completed_.load() < server_.options().drainTime().count());
-  ++drain_time_completed_;
+  const auto remaining_time = std::chrono::duration_cast<std::chrono::seconds>(
+      drain_deadline_ - server_.dispatcher().timeSource().monotonicTime());
+  const auto elapsed_time = server_.options().drainTime() - remaining_time;
 
-  if (drain_time_completed_.load() < server_.options().drainTime().count()) {
-    drain_tick_timer_->enableTimer(std::chrono::milliseconds(1000));
-  } else if (drain_sequence_completion_) {
-    drain_sequence_completion_();
-  }
+  // P(shut down connections) = elapsed_drain_time / overall_drain_timeout
+  return static_cast<uint64_t>(elapsed_time.count()) >
+         (server_.random().random() % server_.options().drainTime().count());
 }
 
-void DrainManagerImpl::startDrainSequence(std::function<void()> completion) {
-  drain_sequence_completion_ = completion;
+void DrainManagerImpl::startDrainSequence(std::function<void()> drain_complete_cb) {
+  drain_complete_cb_ = drain_complete_cb;
   ASSERT(!draining_);
   ASSERT(!drain_tick_timer_);
   draining_ = true;
-  drain_tick_timer_ = server_.dispatcher().createTimer([this]() -> void { drainSequenceTick(); });
-  drainSequenceTick();
+  drain_tick_timer_ =
+      server_.dispatcher().createTimer(drain_complete_cb_ ? drain_complete_cb_ : [] {});
```

Comment:
is there a prod scenario where drain_complete_cb is false? I see it in the tests, but you could just put the empty callback there, rather than adding to the complexity of the prod code.

Reply:
The problem is actually with the timer API, which has an assertion on callbacks being non-false. I don't really think the check makes sense, since noop functions fail the assert but don't do anything when called, but I didn't think it made sense to mess with that in this PR.

Reply:
That's fine. Just assert that drain_callback_cb_ is truthy, and change the test to pass an empty callback rather than null.

Reply:
To answer your first question, there actually is a prod scenario. server.cc:139 has a `drain_manager_->startDrainSequence(nullptr);`. I can just change that though?

The reason this is coming up now is because prior, the timer didn't receive the cb directly, but rather a lambda that invoked the cb. So I guess another option that's neater is `server_.dispatcher().createTimer([]() { drain_complete_cb(); }`.

Reply:
Yeah I think we should just change the callsite that passes null for the functor to simplify all the other code paths.

Reply:
Done.